### PR TITLE
fix(marketo): unhandled exception status-code fix

### PR DESCRIPTION
**Pull request by sanpj2292**
## Description of the change

https://www.notion.so/rudderstacks/subteam-S03SHJ20350-integration-on-call-we-are-failing-to-unmarshal-some-transformer-s-responses-in-b7eee3c0393847a6b6d22fb9411924d7

## Type of change
- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/marketo/util.js [hunk 1]
```diff
@@ -91,14 +99,13 @@ const marketoResponseHandler = (destResponse, sourceMessage, stage) => {
   }
   // non 2xx failure
   throw new ErrorBuilder()
-    .setStatus(status)
+    .setStatus(400)
```
Comment (ItsSudip) [inline v0/destinations/marketo/util.js:102]: Why we are making it to 400 as we can send whatever response we are getting from the destination?

File: v0/destinations/marketo/util.js [hunk 2]
```diff
@@ -13,7 +13,15 @@ const { TRANSFORMER_METRIC } = require("../../util/constant");
 const ErrorBuilder = require("../../util/error");
 
 const MARKETO_RETRYABLE_CODES = ["601", "602", "604", "611"];
-const MARKETO_ABORTABLE_CODES = ["600", "603", "605", "609", "610", "612"];
+const MARKETO_ABORTABLE_CODES = [
```
Comment (ItsSudip) [inline v0/destinations/marketo/util.js:16]: Can we add some documentation links from the destination for these error types if available?

File: v0/destinations/marketo/util.js (hunk 1 above)
Comment (sanpj2292) [reply v0/destinations/marketo/util.js:102]: We are getting `200` status from destination. Any successful should've already been returned from **[here](https://github.com/rudderlabs/rudder-transformer/pull/1432/files#diff-6d523239215a17c74a801bf8be57c380ebdc658bd1a763c98a2faf9e76021e8fR92-R93)**

So it doesn't make sense to send the status as is. Let me know if it makes sense

File: v0/destinations/marketo/util.js (hunk 2 above)
Comment (sanpj2292) [reply v0/destinations/marketo/util.js:16]: Sure

Comment (sanpj2292) [reply v0/destinations/marketo/util.js:21]: Added the doc. Please check

Comment (ItsSudip) [reply v0/destinations/marketo/util.js:21]: You provided the wrong link. Can you check if [this](https://developers.marketo.com/rest-api/error-codes/) link is suitable?

Comment (sanpj2292) [reply v0/destinations/marketo/util.js:21]: Sorry. Corrected the link now.

**Review by ItsSudip — APPROVED** [review]

File: v0/destinations/marketo/util.js [hunk 3]
```diff
@@ -91,14 +104,13 @@ const marketoResponseHandler = (destResponse, sourceMessage, stage) => {
...
-    .setStatus(status)
+    .setStatus(400)
     .setMessage(`Error occured ${sourceMessage}`)
     .setDestinationResponse(destResponse)
     .setStatTags({
       destType: DESTINATION,
       stage,
-      scope: TRANSFORMER_METRIC.MEASUREMENT_TYPE.API.SCOPE,
-      meta: getDynamicMeta(status)
+      scope: TRANSFORMER_METRIC.MEASUREMENT_TYPE.EXCEPTION.SCOPE
```
Comment (saikumarrs) [inline v0/destinations/marketo/util.js:113]: Why is it an exception scope?

File: v0/destinations/marketo/util.js [hunk 4]
```diff
@@ -91,14 +104,13 @@ const marketoResponseHandler = (destResponse, sourceMessage, stage) => {
   }
   // non 2xx failure
   throw new ErrorBuilder()
-    .setStatus(status)
+    .setStatus(400)
```
Comment (saikumarrs) [inline v0/destinations/marketo/util.js:107]: Let's set the stage as well if it's not defaulted already.

File: v0/destinations/marketo/util.js [hunk 5]
```diff
@@ -154,6 +166,20 @@ const responseHandler = (destinationResponse, _dest) => {
...
+      .setStatus(400)
+      .setMessage(`Error occured during Marketo Response Handling`)
+      .setDestinationResponse(destinationResponse.response)
+      .setStatTags({
+        destType: DESTINATION,
+        stage: TRANSFORMER_METRIC.TRANSFORMER_STAGE.RESPONSE_TRANSFORM,
+        scope: TRANSFORMER_METRIC.MEASUREMENT_TYPE.EXCEPTION.SCOPE
+      })
+      .build();
+  }
```
Comment (saikumarrs) [inline v0/destinations/marketo/util.js:182]: Should move this to `marketoApplicationErrorHandler`.

Comment (saikumarrs) [reply v0/destinations/marketo/util.js:182]: or rather `marketoResponseHandler`.

File: v0/destinations/marketo/util.js [hunk 6]
```diff
@@ -154,6 +166,20 @@ const responseHandler = (destinationResponse, _dest) => {
     "during Marketo Response Handling",
```
Comment (saikumarrs) [inline v0/destinations/marketo/util.js:166]: Can we refactor `responseHandler` and `marketoResponseHandler`? Both do the same but for different stages and only slight changes in implementation.

File: v0/destinations/marketo/util.js (hunk 3 above)
Comment (utsabc) [reply v0/destinations/marketo/util.js:113]: Later on it would be difficult to debug if we set API scope

**Review by utsabc — APPROVED** [review]

File: v0/destinations/marketo/util.js (hunk 3 above)
Comment (saikumarrs) [reply v0/destinations/marketo/util.js:113]: Can you please clarify?

File: v0/destinations/marketo/util.js (hunk 4 above)
Comment (sanpj2292) [reply v0/destinations/marketo/util.js:107]: The stage property is being sent as an argument and that is being set as part of `setStatTags`

File: v0/destinations/marketo/util.js (hunk 3 above)
Comment (sanpj2292) [reply v0/destinations/marketo/util.js:113]: This is done for two things
- To indicate that the response sent is one of the unhandled status-code cases
- To get ourselves notified(as `exception` is highest priority) saying "this is an unhandled case, please handle it"

Let me know if it makes sense

File: v0/destinations/marketo/util.js (hunk 5 above)
Comment (sanpj2292) [reply v0/destinations/marketo/util.js]: Moved to `marketoResponseHandler`. Can you please check ?

File: v0/destinations/marketo/util.js (hunk 6 above)
Comment (sanpj2292) [reply v0/destinations/marketo/util.js:171]: Have refactored it to use only one responseHandler. Can you please check and let me know if it is ok ?

File: v0/destinations/marketo/util.js (hunk 3 above)
Comment (saikumarrs) [reply v0/destinations/marketo/util.js:135]: We are misusing the scopes. If that's a workaround for now, leave a comment to refactor later.

**Review by saikumarrs — DISMISSED** [review]

File: v0/destinations/marketo/util.js (hunk 3 above)
Comment (sanpj2292) [reply v0/destinations/marketo/util.js:135]: I don't think we are mis-using the scopes at all. It calls out for an unhandled error case in terms of response from destination. But if you feel strongly about me mis-using the scope, will change it

Comment (sanpj2292) [reply v0/destinations/marketo/util.js:135]: The changes have been, can you check it and let me know if the scope is properly set ?

**Review by saikumarrs — APPROVED** [review]

**Review by ItsSudip — APPROVED** [review]

**Pull request merged**